drivers/pktmem.c

```diff
@@ -97,86 +176,102 @@ static void net_freeDMAPbuf(struct pbuf *p)
 	SYS_ARCH_DECL_PROTECT(old_level);
 
 	SYS_ARCH_PROTECT(old_level);
-	net_freePktBuf(p->payload);
+	struct pktmem_pbuf_custom *ppc = (void *)p;
+	net_freePktBuf(p->payload, ppc->pktmem_start);
 	SYS_ARCH_UNPROTECT(old_level);
 
-	mem_free(p);
+	mem_free(ppc);
 }
 
 
-static ssize_t net_allocPktBuf(void **bufp)
+static ssize_t net_allocPktBuf(void **bufp, void **pktmem_start)
 {
 	SYS_ARCH_DECL_PROTECT(old_level);
-	buf_page_head_t *ph;
+	pktmem_info_t *info;
 	unsigned i;
 
 	SYS_ARCH_PROTECT(old_level);
-	if (pkt_bufs_free == 0) {
+	if (pkt_bufs_free == 0 || pktmem_lru.next == &pktmem_lru) {
 		SYS_ARCH_UNPROTECT(old_level);
 
-		ph = dmammap(PAGE_SIZE);
-		if (ph == NULL) {
+		info = dmammap(pktmem_opts.alloc_sz);
+		if (info == NULL) {
 			printf("mmap: no memory?\n");
 			return 0;
 		}
+		info = (void *)((uintptr_t)info + pktmem_opts.pktmem_info_offset);
 
-		memset(ph, 0, CACHE_LINE_SIZE);
-		ph->free_mask = (1 << PKT_BUF_CNT) - 1;
+		memset(info, 0, sizeof(*info));
+		info->free_mask = pktmem_opts.free_mask_full;
 
 		SYS_ARCH_PROTECT(old_level);
-		net_listAdd(ph, &pkt_buf_lru);
-		pkt_bufs_free += PKT_BUF_CNT;
+		net_listAdd(info, &pktmem_lru);
+		pkt_bufs_free += pktmem_opts.pkt_buf_cnt;
 	}
 	else {
-		ph = pkt_buf_lru.next;
+		info = pktmem_lru.next;
 	}
 
-	i = __builtin_ctz(ph->free_mask);
 	--pkt_bufs_free;
-	ph->free_mask &= ~(1 << i);
-	if (ph->free_mask == 0) {
-		net_listDel(ph);
+
+	i = __builtin_ctz(info->free_mask);
```

Comment:
You could optimize the new pktmem case by moving this code into `else` block. In the case of a newly allocated pktmem with only one packet buffer, we don't need to add and remove it from the list.